### PR TITLE
[JFG-1035] Upload new tests to the running Jagger Master

### PR DESCRIPTION
**Pull request by a-badaev**

### CODE REVIEW

File: jaas/src/main/java/com/griddynamics/jagger/jaas/service/impl/TestEnvironmentServiceImpl.java [hunk 1]
```diff
@@ -100,7 +94,7 @@ else if (!isEqualCollection(testEnvToUpdate.getLoadScenarios(), newTestEnv.getLo
         }
 
         if (testEnvToUpdate.getRunningLoadScenario() != newTestEnv.getRunningLoadScenario()) {
-            testEnvToUpdate.setRunningLoadScenario(getNewRunningLoadScenario(newTestEnv, testEnvToUpdate));
+            testEnvToUpdate.setRunningLoadScenario(newTestEnv.getRunningLoadScenario());
```
Comment (anton-antonenko) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/service/impl/TestEnvironmentServiceImpl.java]: This code may produce persistence exceptions if newTestEnv.getRunningLoadScenario() is not persisted yet.
To omit such situations I'd propose you to check if env has this load scenario and if no - add it.

Comment (anton-antonenko) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/service/impl/TestEnvironmentServiceImpl.java]: Something like this. But, if I'm not mistaken, you cannot add new load scenario to test env and set it as runing in one transaction... so, maybe, you'll need to make 2 updates

`
  if (!testEnvToUpdate.getLoadScenarios().contains(newTestEnv.getRunningLoadScenario()))
            testEnvToUpdate.getLoadScenarios().add(newTestEnv.getRunningLoadScenario());
`

Comment (a-badaev) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/service/impl/TestEnvironmentServiceImpl.java]: do you mean I can't update more then 1 field in a single transaction? or what is it related with?

Comment (anton-antonenko) [inline jaas/src/main/java/com/griddynamics/jagger/jaas/service/impl/TestEnvironmentServiceImpl.java]: No. It's related to LoadScenarioEntity table. When you add new load scenario to test env and update it, this new load scenarion is persisted to db and obtains its id.
But when you set runningLoadScenario in the same transaction it's not persisted yet and exception will occur due to this. That's why you need to persist load scenario firstly by adding it to env and only after that you'll be able to set runningLoadScenario.

File: chassis/coordinator/src/main/java/com/griddynamics/jagger/coordinator/DefaultNodeContext.java [hunk 2]
```diff
@@ -38,7 +38,13 @@ public NodeId getId() {
 
     @Override
     public <T> T getService(Class<T> clazz) {
-        return (T) services.get(clazz);
+        if (services.containsKey(clazz)) {
+            return (T) services.get(clazz);
+        }
+        try {
+            return clazz.newInstance();
+        } catch (Exception e) {}
```
Comment (dlatnikov) [inline chassis/coordinator/src/main/java/com/griddynamics/jagger/coordinator/DefaultNodeContext.java]: why silent catch?

File: chassis/core/src/main/java/com/griddynamics/jagger/JaggerLauncher.java [hunk 3]
```diff
@@ -179,9 +180,19 @@ public void run() {
                     )) {
                         masterToJaasCoordinator.register();
                         while (masterToJaasCoordinator.isStandBy()) {
-                            environmentProperties
-                                    .setProperty(TEST_CONFIG_NAME_PROP, masterToJaasCoordinator.awaitConfigToExecute());
-                            doLaunchMaster(directory);
+                            MasterToJaasCoordinator.JaasResponse jaasResponse =
+                                    masterToJaasCoordinator.awaitConfigToExecute();
+                            if (jaasResponse.getTestProjectUrl() == null) {
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/JaggerLauncher.java]: please add comment that in this case we are running existing load scenario

File: chassis/core/src/main/java/com/griddynamics/jagger/JaggerLauncher.java [hunk 4]
```diff
@@ -179,9 +180,19 @@ public void run() {
...
+                                environmentProperties
+                                        .setProperty(LOAD_SCENARIO_ID_PROP, jaasResponse.getLoadScenarioName());
+                                System.setProperty(
+                                        USER_CONFIGS_PACKAGE, environmentProperties.getProperty(USER_CONFIGS_PACKAGE));
+                                doLaunchMaster(directory);
+                            } else {
+                                // "file:///Users/abadaev/.m2/repository/org/test/load-scenario/1.0-SNAPSHOT/load-scenario-1.0-SNAPSHOT.jar"
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/JaggerLauncher.java]: comment

File: chassis/core/src/main/java/com/griddynamics/jagger/JaggerLauncher.java [hunk 5]
```diff
@@ -195,6 +205,42 @@ public void run() {
         builder.addMainTask(masterTask);
     }
     
+    private static void executeDynamicLoadScenario(String loadScenarioId, String customClassesUrl, URL directory) {
+    
+        environmentProperties.setProperty(LOAD_SCENARIO_ID_PROP, loadScenarioId);
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/JaggerLauncher.java]: loadScenrioId can be null at this point. is it ok?

File: chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java [hunk 6]
```diff
@@ -288,39 +290,46 @@ private void tryToOfferNextConfigToExecute(ResponseEntity<String> responseEntity
             if (requestEntity.getBody().getStatus() != TestEnvironmentStatus.PENDING) {
                 return;
             }
-            String configName = getNextConfigToExecute(responseEntity);
-            if (configName == null) {
+            String loadScenarioName = extractHeader(responseEntity, TestEnvUtils.LOAD_SCENARIO_HEADER);
+            String testProjectUrl = extractHeader(responseEntity, TestEnvUtils.TEST_PROJECT_URL_HEADER);
+            if (loadScenarioName == null && testProjectUrl == null) {
```
Comment (dlatnikov) [inline chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java]: are you sure about \&&

File: chassis/core/src/main/java/com/griddynamics/jagger/JaggerLauncher.java (hunk 5 above)
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/JaggerLauncher.java]: yes. it is handled later

File: chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java (hunk 6 above)
Comment (a-badaev) [inline chassis/core/src/main/java/com/griddynamics/jagger/master/MasterToJaasCoordinator.java]: yes

**Pull request merged**